wom/src/main/scala/wom/values/WomFile.scala

```diff
@@ -1,67 +1,154 @@
+  // TODO: WOM: WOMFILE: When WDL supports the other parts beyond `value`, toWomString should return the _WDL_ string.
+  override def toWomString = s""""$value""""
+
+  override def mapFile(f: String => String): WomSingleDirectoryOrFile
+}
 
-  def mapSingleFile(womFile: WomSingleFile, f: String => String): WomSingleFile = {
-    WomSingleFile(f(womFile.value))
+/**
+  * A directory with an optional listing of other files/directories.
+  *
+  * A user may supply a directory listing where the files/directories are not technically sub-directories of this
+  * parent.
+  *
+  * Directories with listings set to None should not be passed into command line generation. Instead, the execution
+  * engine should create a copy of this instance by locating the files/directories within the `value` and filling in the
+  * listing with Some Seq.
+  *
+  * @param value         The location of the directory, possibly in the cloud.
+  * @param listingOption An optional listing of files/directories, either supplied by a user or generated by the engine.
+  */
+case class WomSingleDirectory(value: String,
+                              listingOption: Option[Seq[WomSingleDirectoryOrFile]] = None
+                             ) extends WomSingleDirectoryOrFile {
+  override val womType: WomType = WomSingleDirectoryType
+
+  override def add(rhs: WomValue): Try[WomValue] = rhs match {
+    case r: WomString => Success(WomSingleDirectory(value + r.value))
+    case r: WomOptionalValue => evaluateIfDefined("+", r, add)
+    case _ => invalid(s"$value + $rhs")
   }
 
-  def mapGlobFile(womFile: WomGlobFile, f: String => String): WomGlobFile = {
-    WomGlobFile(f(womFile.value))
+  override def mapFile(f: String => String): WomSingleDirectory = {
+    this.copy(value = f(value), listingOption.map(_.map(_.mapFile(f))))
   }
 }
 
-case class WomSingleFile(value: String) extends WomFile {
-  override val womType: WomType = WomSingleFileType
+/**
+  * A file with optional sub files/directories. The contents may have been provided by the user and will need to be
+  * written by the engine before executing the command.
+  *
+  * @param value          The location of the file, possibly in the cloud.
+  * @param checksumOption An optional checksum of the file contents.
+  * @param sizeOption     An optional size of the file contents in bytes.
+  * @param formatOption   An optional format description of the file contents.
+  * @param contentsOption The optional text contents of the file.
+  * @param secondaryFiles Any files associated with this file.
+  */
+case class WomSingleFile(value: String,
+                         checksumOption: Option[String] = None,
+                         sizeOption: Option[Long] = None,
+                         formatOption: Option[String] = None,
+                         contentsOption: Option[String] = None,
+                         secondaryFiles: Seq[WomSingleDirectoryOrFile] = Vector.empty
```

Comment:
ToL: Is it just me or is `secondaryFiles` an odd field for `WomSingleFile` to have? ie we can no longer make assumptions like "this is a single file"

Reply:
Yeah I agree this feels a bit weird but I don't have a better suggestion